pyhaystack/client/mixins/vendor/niagara/bql.py

```diff
@@ -0,0 +1,95 @@
+#!python
+# -*- coding: utf-8 -*-
+"""
+Niagara allows BQL requests to browse the database
+
+This is a BETA function that will need to be tested by users
+
+"""
+
+from requests.compat import urljoin
+import io
+import pandas as pd
+
+from ....ops.grid import BaseAuthOperation
+from .....util.asyncexc import AsynchronousException
+
+class BQLOperation(BaseAuthOperation):
+       
+    def __init__(self, session, bql, args=None, **kwargs):
+        """
+        Initialise a GET request for the BQL with the given request and arguments.
+
+        :param session: Haystack HTTP session object.
+        :param bql: BQL Request 
+        :param args: Dictionary of key-value pairs to be given as arguments.
+        """
+        self._log = session._log.getChild('bql.%s' % bql)
+        bql_request = 'ord?' + bql + '%7Cview:file:ITableToCsv'

```

Comment:
You might want to have a look at `urllib.parse.quote` as a more robust way to quote fields in URIs:

```
Help on function quote in urllib.parse:

urllib.parse.quote = quote(string, safe='/', encoding=None, errors=None)
    quote('abc def') -> 'abc%20def'
    
    Each part of a URL, e.g. the path info, the query, etc., has a
    different set of reserved characters that must be quoted.
    
    RFC 2396 Uniform Resource Identifiers (URI): Generic Syntax lists
    the following reserved characters.
    
    reserved    = ";" | "/" | "?" | ":" | "@" | "&" | "=" | "+" |
                  "$" | ","
    
    Each of these characters is reserved in some component of a URL,
    but not necessarily in all of them.
    
    By default, the quote function is intended for quoting the path
    section of a URL.  Thus, it will not encode '/'.  This character
    is reserved, but in typical usage the quote function is being
    called on a path where the existing slash characters are used as
    reserved characters.
    
    string and safe may be either str or bytes objects. encoding and errors
    must not be specified if string is a bytes object.
    
    The optional encoding and errors parameters specify how to deal with
    non-ASCII characters, as accepted by the str.encode method.
    By default, encoding='utf-8' (characters are encoded with UTF-8), and
    errors='strict' (unsupported characters raise a UnicodeEncodeError).
```

In Python2, this is `urllib.quote`; so to use it:

```
try:
    from urllib.parse import quote as quote_uri
except ImportError:
    # Python 2.7 dinosaur
    from urllib import quote as quote_uri
```

then

```
bql_request = 'ord?' + quote_uri(bql) + '%7Cview:file:ITableToCsv'
```

Another function that might work too, is `quote_plus`, from the same library; which replaces `' '` with `'+'` instead of `'%20'`.

Reply:
Nice ! I'll stick with `%20` because I know it works with Niagara..... I don't know for `+`

Reply:
If you need `/` escaped, try this:
```
In [4]: quote_uri('/ foo /', safe='')
Out[4]: '%2F%20foo%20%2F'
```